hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/MetaBrowser.java

```diff
@@ -0,0 +1,378 @@
+      new QueryStringEncoder(request.getRequestURI());
+    addParam(encoder, NAME_PARAM, name);
+    addParam(encoder, SCAN_LIMIT_PARAM, scanLimit);
+    addParam(encoder, SCAN_REGION_STATE_PARAM, scanRegionState);
+    addParam(encoder, SCAN_TABLE_PARAM, scanTable);
+    return encoder;
+  }
+
+  public String buildFirstPageUrl() {
+    return buildFirstPageEncoder().toString();
+  }
+
+  public static String buildStartParamFrom(final byte[] lastRow) {
+    if (lastRow == null) {
+      return null;
+    }
+    return urlEncode(Bytes.toStringBinary(lastRow));
+  }
+
+  public String buildNextPageUrl(final byte[] lastRow) {
+    final QueryStringEncoder encoder = buildFirstPageEncoder();
+    final String startRow = buildStartParamFrom(lastRow);
+    addParam(encoder, SCAN_START_PARAM, startRow);
+    return encoder.toString();
+  }
+
+  private static String urlEncode(final String val) {
+    if (StringUtils.isEmpty(val)) {
+      return null;
+    }
+    try {
+      return URLEncoder.encode(val, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException e) {
+      return null;
+    }
+  }
+
+  private static String urlDecode(final String val) {
+    if (StringUtils.isEmpty(val)) {
+      return null;
+    }
+    try {
+      return URLDecoder.decode(val, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException e) {
+      return null;
+    }
+  }
+
+  private static Integer tryParseInt(final String val) {
+    if (StringUtils.isEmpty(val)) {
+      return null;
+    }
+    try {
+      return Integer.parseInt(val);
+    } catch (NumberFormatException e) {
+      return null;
+    }
+  }
+
+  private static <T extends Enum<T>> T tryValueOf(final Class<T> clazz,
+    final String value) {
+    if (clazz == null || value == null) {
+      return null;
+    }
+    try {
+      return T.valueOf(clazz, value);
+    } catch (IllegalArgumentException e) {
+      return null;
+    }
+  }
+
+  private static String buildScanLimitExceededErrorMessage(final int requestValue) {
+    return String.format(
+      "Requested SCAN_LIMIT value %d exceeds maximum value %d.", requestValue, SCAN_LIMIT_MAX);
+  }
+
+  private static String buildScanLimitMalformedErrorMessage(final String requestValue) {
+    return String.format(
+      "Requested SCAN_LIMIT value '%s' cannot be parsed as an integer.", requestValue);
+  }
+
+  private static String buildScanLimitLTEZero(final int requestValue) {
+    return String.format("Requested SCAN_LIMIT value %d is <= 0.", requestValue);
+  }
+
+  private static String buildScanRegionStateMalformedErrorMessage(final String requestValue) {
+    return String.format(
+      "Requested SCAN_REGION_STATE value '%s' cannot be parsed as a RegionState.", requestValue);
+  }
+
+  /**
+   * Encapsulates the results produced by this {@link MetaBrowser} instance.
+   */
+  public final class Results implements AutoCloseable, Iterable<RegionReplicaInfo> {
```

Comment:
I was about to suggest something like this.. wrap the output in some auto-closeable implementation that can be used in callers.. neat...